### PR TITLE
Refactor accumulator

### PR DESCRIPTION
**Pull request by 0xpause**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #3390 

## What is the new behavior?

- Change AccumulatorStorage from key-value `(hash, node)` to `(index, hash)`
- Add StorageVersion::V4 and db_upgrade from v3 to v4

### CODE REVIEW

**Comment by nkysg** [comment]
@0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。

File: commons/accumulator/src/node_index.rs [hunk 1]
```diff
@@ -19,6 +21,15 @@ pub enum NodeDirection {
 }
 
 impl NodeIndex {
+    pub fn encode(self) -> Result<Vec<u8>> {
+        Ok(self.0.to_be_bytes().to_vec())
+    }
+
+    #[allow(clippy::redundant_slicing)]
+    pub fn decode(data: &[u8]) -> Result<Self> {
+        Ok(Self((&data[..]).read_u64::<BigEndian>().unwrap()))
```
Comment (nkysg) [inline commons/accumulator/src/node_index.rs:30]: 这里不要用unwrap,非法输入会panic, 用?

File: commons/accumulator/src/tree.rs [hunk 2]
```diff
@@ -263,71 +252,12 @@ impl AccumulatorTree {
...
-                    error!(
-                        "can not find accumulator node by hash :{:?} in store: {:?}",
-                        temp_node_hash,
-                        self.store.store_type()
-                    );
-                }
-            }
-        }
-        bail!("node hash not found:{:?}", index)
+        Ok(self.store.get_node(index)
```
Comment (nkysg) [inline commons/accumulator/src/tree.rs:258]: 这里的逻辑不正确，你存储的只是frozen的NodeIndex的节点,对于非Frozen的NodeIndex节点, 并没有存储到RocksDB内，所以这里非Frozen的NodeIndex的节点应该计算出来。你这里可能只是你测试中从Cache中能拿到数据，但这个Cache是个LruCache，实际中可能被Cache淘汰掉

File: commons/accumulator/src/tree.rs [hunk 3]
```diff
@@ -263,71 +252,12 @@ impl AccumulatorTree {
...
+            .unwrap_or(Some(*ACCUMULATOR_PLACEHOLDER_HASH))
+            .unwrap())
```
Comment (nkysg) [inline commons/accumulator/src/tree.rs:260]: 这里不要用unwrap,写测试代码可以用，写生产代码不要使用，可以看看项目中使用，比如ok_or_else

File: commons/accumulator/src/tree.rs (hunk 2 above)
Comment (nkysg) [reply commons/accumulator/src/tree.rs:258]: 针对你存储的只是Frozen的NodeIndex的节点，一个节点刚开始是NotFrozen的你读取它的HashValue, 添加Leaves后变成Frozen了，这个测试代码中也需要添加下

File: commons/accumulator/src/accumulator_test.rs [hunk 4]
```diff
@@ -252,8 +240,9 @@ fn test_flush() {
     let _root_hash = accumulator.append(&leaves).unwrap();
     accumulator.flush().unwrap();
     //get from storage
-    for node_hash in leaves {
-        let node = mock_store.get_node(node_hash).unwrap();
+    for leaf_index in 0..20 {
+        let index = NodeIndex::from_leaf_index(leaf_index);
+        let node = mock_store.get_node(index).unwrap();
```
Comment (nkysg) [inline commons/accumulator/src/accumulator_test.rs:245]: 这里get_node需要添加新的测试代码，描述见tree.rs

File: storage/src/upgrade.rs [hunk 5]
```diff
@@ -151,6 +156,60 @@ impl DBUpgrade {
         Ok(())
     }
 
+    fn db_upgrade_v3_v4(instance: &mut StorageInstance) -> Result<()> {
```
Comment (nkysg) [inline storage/src/upgrade.rs:159]: 这里upgrade 基于分叉可能会有问题，我想下怎么描述，以及是否需要添加新的测试代码。

**Comment by 0xpause** [comment]
> @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。

I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
```
test test_block_chain_txn_info_fork_mapping ... FAILED
test test_find_ancestor_fork ... FAILED
test test_switch_epoch ... FAILED
test test_uncle ... FAILED
test test_uncle_exist ... FAILED
```
I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318 . And I think this is a serious problem that prevents continue refactoring.

The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.

**Comment by nkysg** [comment]
> test_uncle



> > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> 
> I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> 
> ```
> test test_block_chain_txn_info_fork_mapping ... FAILED
> test test_find_ancestor_fork ... FAILED
> test test_switch_epoch ... FAILED
> test test_uncle ... FAILED
> test test_uncle_exist ... FAILED
> ```
> 
> I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> 
> The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.

原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。

**Comment by 0xpause** [comment]
> > test_uncle
> 
> > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > 
> > 
> > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > ```
> > test test_block_chain_txn_info_fork_mapping ... FAILED
> > test test_find_ancestor_fork ... FAILED
> > test test_switch_epoch ... FAILED
> > test test_uncle ... FAILED
> > test test_uncle_exist ... FAILED
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> 
> 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。

"产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?

**Comment by nkysg** [comment]
> > > test_uncle
> > 
> > 
> > > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > > 
> > > 
> > > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > > ```
> > > test test_block_chain_txn_info_fork_mapping ... FAILED
> > > test test_find_ancestor_fork ... FAILED
> > > test test_switch_epoch ... FAILED
> > > test test_uncle ... FAILED
> > > test test_uncle_exist ... FAILED
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> > 
> > 
> > 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。
> 
> "产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?

这里的Node是指运行starcoin的客户端。这里表述错了，产生分叉是对两个客户端来说的，POW共识的永远选最长的链。这里有POW consensus 这点说明
> > >
Miners are incentivised to do this work on the main Ethereum chain. There is little incentive for a subset of miners to start their own chain – it undermines the system. Blockchains rely on having a single state as a source of truth. And users will always choose the longest or "heaviest" chain.
> > >
https://ethereum.org/en/developers/docs/consensus-mechanisms/pow/#security

**Comment by nkysg** [comment]
> > > test_uncle
> > 
> > 
> > > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > > 
> > > 
> > > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > > ```
> > > test test_block_chain_txn_info_fork_mapping ... FAILED
> > > test test_find_ancestor_fork ... FAILED
> > > test test_switch_epoch ... FAILED
> > > test test_uncle ... FAILED
> > > test test_uncle_exist ... FAILED
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> > 
> > 
> > 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。
> 
> "产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?

这里的Node是指运行starcoin的客户端。这里表述错了，产生分叉是对两个客户端来说的，POW共识的永远选最长的链。这里有POW consensus 这点说明
> > >
Miners are incentivised to do this work on the main Ethereum chain. There is little incentive for a subset of miners to start their own chain – it undermines the system. Blockchains rely on having a single state as a source of truth. And users will always choose the longest or "heaviest" chain.
> > >
https://ethereum.org/en/developers/docs/consensus-mechanisms/pow/#security

File: commons/accumulator/src/tree.rs (hunk 2 above)
Comment (0xpause) [reply commons/accumulator/src/tree.rs:258]: 1. 从[这里](https://github.com/starcoinorg/starcoin/blob/1c4190b922592ba7f244f08640e6933a80baccfe/commons/accumulator/src/tree.rs#L165)可以看出，Frozen和NotFrozen节点都会被写入RocksDB中。通过运行代码得到确认：append 0,1,2 三个叶子节点，将会写入index 0-5的6个节点，其中3，5是NotFrozen的。事实上，只有 placeholder 节点不会被写入RocksDB，也不会写入 Cache 中。
2. 因此，这里的逻辑是：凡是RocksDB中不存在的index，一律按照 placeholder 处理。我认为这样处理是没有问题的，如果确实需要知道查询的 index 是否存在，那也可以检查，只有存在 sibling 的 placeholder 节点才返回 hash， 其余的按 index 未找到处理。
3. 由于 Frozen 节点不会再更新，NotFrozen 节点变成 Frozen 状态后，曾经的 NotFrozen 的 Hash 被 Frozen 的 Hash 覆盖，而且 NotFrozen 的 Hash 不再有消费者；另外，从代码上看，也没有其他地方会消费 is_frozen 状态，因此我认为 is_frozen 这个状态标志也是不需要的。

Comment (nkysg) [reply commons/accumulator/src/tree.rs:258]: 1.你给出这个链接https://github.com/starcoinorg/starcoin/blob/1c4190b922592ba7f244f08640e6933a80baccfe/commons/accumulator/src/tree.rs#L165 是基于原来(Hash(Node), encode(Node))的存储方式，我说的是这里 https://github.com/0xpause/starcoin/blob/accumulator_storage/storage/src/upgrade.rs#L169
你在将原来的(Hash(Node), encode(Node))升级为(Index,Hash(Node))只存储了is_frozen的Node。
cache是这个https://github.com/0xpause/starcoin/blob/accumulator_storage/commons/accumulator/src/tree.rs#L27
节点计算root hash值时候会更新index_cache
2. 你修改为(Index, Hash(Node))时候你的存储是将Frozen和NotFrozen都存储都写入KvStore了，如果原来是NotFrozen的新添加节点后可能会变成Frozen,这样有几个问题
1) NotFrozen的变成Frozen，如果NotFrozen的写入磁盘，变更后再写入磁盘，实际上写入了两次，写存储是比较耗时的操作比内存计算慢很多，rust和C++这种级别的内存计算可以每秒钟千万个级别，存储大概慢1000级别
2) 在基于proof的时候会读取某个历史状态，存储在磁盘中是Frozen的，proof中是NoFrozen的，你通过KvStore获取是错误的值，这个的计算出来。

Comment (0xpause) [reply commons/accumulator/src/tree.rs:258]: 1. 如前所述，master分支 append 新节点并 flush 后，会将 Frozen 和 NotFrozen 的节点都写入磁盘中。因此，old storage 中一定存在 NotFrozen 的节点；
2. db_upgrade 时并非只升级了 Frozen 节点。https://github.com/0xpause/starcoin/blob/accumulator_storage/storage/src/upgrade.rs#L169 这里的逻辑是：对于某个存在的 NodeIndex， old storage 中可能存在两种情况：a) 有一个 Frozen 节点 x  和一个 NotFrozen 节点  y；b）只有一个 NotFrozen 节点。
    - 对于情况a），由于不能保证节点 x，y 被遍历的先后顺序，因此添加 Node 时检查new storage，如果 NodeIndex 不存在new storage中，那么直接写入即可；如果 NodeIndex 已经存在，那么有两种情况：
         - 如果新添加的 Node 是 Frozen 的，说明之前添加过一个 NotFrozen 的节点，现在只需将 Frozen 节点写入覆盖即可；
         - 如果新添加的 Node 是 NotFrozen的，说明之前添加过一个 Frozen 的节点，那么不做任何处理； 
   - 对于情况b），直接将节点添加到new storage中即可。
  
    由于 old storage 中同时存在 Frozen 和 NotFrozen 的，因此这里升级也不可避免的将 NotFrozen 的节点引入到 new storage 中
3. AccumulatorInfo 只用了 frozen_subtree_roots ，不需要每个节点的 frozen 状态。frozen_subtree_roots 可以通过 num_leaves 推算出来
4. 之前没有考虑写磁盘的性能问题，那这里需要修改一下，在 append 新节点后只将 Frozen 节点写入磁盘，NotFrozen 节点写入 cache.

**Comment by nkysg** [comment]
> > > test_uncle
> > 
> > 
> > > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > > 
> > > 
> > > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > > ```
> > > test test_block_chain_txn_info_fork_mapping ... FAILED
> > > test test_find_ancestor_fork ... FAILED
> > > test test_switch_epoch ... FAILED
> > > test test_uncle ... FAILED
> > > test test_uncle_exist ... FAILED
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> > 
> > 
> > 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。
> 
> "产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?

这个理解我可能理解有误，我再想下

File: storage/src/upgrade.rs (hunk 5 above)
Comment (0xpause) [reply storage/src/upgrade.rs:159]: 1. 回滚不会清除对应的 Node 数据吗？ 
2. 如果回滚没有清除数据， 感觉只从 Storage 里面获取数据没有有效信息来区分一个节点是否是有效的节点。
3. 一个可能的办法是重新构造 MerckleTree ？但这需要 AccumulatorTreeInfo。这样是否可行？

**Comment by 0xpause** [comment]
> > > > test_uncle
> > > 
> > > 
> > > > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > > > 
> > > > 
> > > > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > > > ```
> > > > test test_block_chain_txn_info_fork_mapping ... FAILED
> > > > test test_find_ancestor_fork ... FAILED
> > > > test test_switch_epoch ... FAILED
> > > > test test_uncle ... FAILED
> > > > test test_uncle_exist ... FAILED
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > > > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> > > 
> > > 
> > > 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。
> > 
> > 
> > "产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?
> 
> 这个理解我可能理解有误，我再想下

你说的是哪部分“理解有误“？

我尝试了“复制数据去生成新的Accumulator2“，原本想直接复制底层 StorageInstance，但这比较困难，原因1) BlockChain 使用的 Store trait 对象，要获取底层 StorageInstance 需要层层暴露太多接口；2）StorageInstance 底层复制也比较麻烦。

我现在写了一个 BlockChain::new_with_uncles_for_test 函数，在里面构造了一个空的 MerkleAccumulator Accumulator2， 然后将 Accumulaor 的 leaves 逐个 append 到新 Accumulator2。 这样可以解决 test_find_ancestor_fork，test_switch_epoch, test_uncle, test_uncle_exist 的问题；但是对于 test_block_chain_txn_info_fork_mapping，需要深入到 OpenedBlock::new](https://github.com/starcoinorg/starcoin/blob/ff3734e8d42f973f5cd767a77e2dd1e681a3d128/chain/open-block/src/lib.rs#L61)  去修改。这也对生产代码产生很大影响，似乎这也不是一条合理的路径。

请问有更好的建议吗？

**Comment by nkysg** [comment]
> > > > > test_uncle
> > > > 
> > > > 
> > > > > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > > > > 
> > > > > 
> > > > > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > > > > ```
> > > > > test test_block_chain_txn_info_fork_mapping ... FAILED
> > > > > test test_find_ancestor_fork ... FAILED
> > > > > test test_switch_epoch ... FAILED
> > > > > test test_uncle ... FAILED
> > > > > test test_uncle_exist ... FAILED
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > > > > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> > > > 
> > > > 
> > > > 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。
> > > 
> > > 
> > > "产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?
> > 
> > 
> > 这个理解我可能理解有误，我再想下
> 
> 你说的是哪部分“理解有误“？
> 
> 我尝试了“复制数据去生成新的Accumulator2“，原本想直接复制底层 StorageInstance，但这比较困难，原因1) BlockChain 使用的 Store trait 对象，要获取底层 StorageInstance 需要层层暴露太多接口；2）StorageInstance 底层复制也比较麻烦。
> 
> 我现在写了一个 BlockChain::new_with_uncles_for_test 函数，在里面构造了一个空的 MerkleAccumulator Accumulator2， 然后将 Accumulaor 的 leaves 逐个 append 到新 Accumulator2。 这样可以解决 test_find_ancestor_fork，test_switch_epoch, test_uncle, test_uncle_exist 的问题；但是对于 test_block_chain_txn_info_fork_mapping，需要深入到 OpenedBlock::new](
> 
> https://github.com/starcoinorg/starcoin/blob/ff3734e8d42f973f5cd767a77e2dd1e681a3d128/chain/open-block/src/lib.rs#L61
> 
> ) 去修改。这也对生产代码产生很大影响，似乎这也不是一条合理的路径。
> 请问有更好的建议吗？

之前说的分叉相关的理解，就是分叉是在一个node的当前block接受到了两个不同的新的block，需求两个block都能执行交易，但是现在这个设计不行，不好意思我搞错了。

**Comment by 0xpause** [comment]
> 



> > > > > > test_uncle
> > > > > 
> > > > > 
> > > > > > > @0xpause please run https://github.com/starcoinorg/starcoin/blob/master/scripts/check_commit.sh, see if there is some warning or errors。
> > > > > > 
> > > > > > 
> > > > > > I have done that, and fix all the fmt check. I also found another problem: all unit tests in Accumulator and Storage passed, however some tests in other modules failed. The failed cases are:
> > > > > > ```
> > > > > > test test_block_chain_txn_info_fork_mapping ... FAILED
> > > > > > test test_find_ancestor_fork ... FAILED
> > > > > > test test_switch_epoch ... FAILED
> > > > > > test test_uncle ... FAILED
> > > > > > test test_uncle_exist ... FAILED
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > I dig into these codes and found that it's caused by multiple AccumulatorTrees as described in [#3390 (comment)](https://github.com/starcoinorg/starcoin/issues/3390#issuecomment-1146609318) . And I think this is a serious problem that prevents continue refactoring.
> > > > > > The reason Accumulator's unit tests(especially test_multiple_chain) passed was because index_cache were used in the tests, instead of the KVStore.
> > > > > 
> > > > > 
> > > > > 原来的Accumlator设计中使用(hash, encode(Node)),所以一个Accumulator分叉成有两个Accumulator1, Accumulator2, 但是新的设计中使用(NodeIndex, Hash(Node)) 一个Accumulator分叉成两个Accumulator1, Accumlator2, Accumlator2会rewrite Accumlator1,最终Accumulator只有一个Accumlator2。产生分叉应该是在两个不同Node上，Accumulator1和Accumulator2应该在两个不同的Node1, Node2上。原先的设计test_uncle这里是用一块内存数据(KvStore)保存了两个不同的Accumulaor,分别是Accumulator1和Accumulator2, 这个新设计测试需要复制数据去生成新的Accumulator2， 而不是在原来基础上直接修改。
> > > > 
> > > > 
> > > > "产生分叉应该是在两个不同Node上"，这里的Node是指运行starcoin的客户端？也就是说一个客户端不会出现分叉的情况，永远只有一个BlockAccumulator和TransactionAccumulator ?
> > > 
> > > 
> > > 这个理解我可能理解有误，我再想下
> > 
> > 
> > 你说的是哪部分“理解有误“？
> > 我尝试了“复制数据去生成新的Accumulator2“，原本想直接复制底层 StorageInstance，但这比较困难，原因1) BlockChain 使用的 Store trait 对象，要获取底层 StorageInstance 需要层层暴露太多接口；2）StorageInstance 底层复制也比较麻烦。
> > 我现在写了一个 BlockChain::new_with_uncles_for_test 函数，在里面构造了一个空的 MerkleAccumulator Accumulator2， 然后将 Accumulaor 的 leaves 逐个 append 到新 Accumulator2。 这样可以解决 test_find_ancestor_fork，test_switch_epoch, test_uncle, test_uncle_exist 的问题；但是对于 test_block_chain_txn_info_fork_mapping，需要深入到 OpenedBlock::new](
> > https://github.com/starcoinorg/starcoin/blob/ff3734e8d42f973f5cd767a77e2dd1e681a3d128/chain/open-block/src/lib.rs#L61
> > 
> > ) 去修改。这也对生产代码产生很大影响，似乎这也不是一条合理的路径。
> > 请问有更好的建议吗？
> 
> 之前说的分叉相关的理解，就是分叉是在一个node的当前block接受到了两个不同的新的block，需求两个block都能执行交易，但是现在这个设计不行，不好意思我搞错了。

你是指 (index, hash) 作为 KVStore 的 key-value 这个方案是行不通吗？

如果是这样，那我就直接把这个 PR 关闭了？

**Comment by nkysg** [comment]
要不改成Draft吧，我再想想看，到时候可以一起讨论下

**Comment by nkysg** [comment]
https://github.com/starcoinorg/starcoin/pull/3551 has improve accumulator append. need improve proof

**Comment by nkysg** [comment]
I wlill close this. Thanks  @0xpause  we only need improve accumlator proof.

**Pull request closed without merging**